Mathlib/Analysis/Normed/Ring/FiniteExtension.lean

```diff
@@ -0,0 +1,261 @@
+  least one power-multiplicative `K`-algebra norm on `L` extending the norm on `K`. This is
+  [BGR, Lemma 3.2.1./3].
+
+## References
+* [S. Bosch, U. Güntzer, R. Remmert, *Non-Archimedean Analysis*][bosch-guntzer-remmert]
+
+## Tags
+
+Basis.norm, nonarchimedean
+-/
+
+noncomputable section
+
+open Finset
+
+section Ring
+
+variable {K L : Type*} [NormedField K] [Ring L] [Algebra K L]
+
+namespace Basis
+
+variable {ι : Type*} [Fintype ι] [Nonempty ι] (B : Basis ι K L)
+
+/-- The function sending an element `x : L` to the maximum of the norms of its coefficients
+  with respect to the `K`-basis `B` of `L`.-/
+def norm : L → ℝ :=
+  fun x ↦ ‖B.repr x (Classical.choose (Finite.exists_max fun i : ι ↦ ‖B.repr x i‖))‖
+
+/-- The norm of a coefficient `x_i` is less than or equal to the norm of `x`. -/
+theorem norm_repr_le_norm {x : L} (i : ι) : ‖B.repr x i‖ ≤ B.norm x :=
+  Classical.choose_spec (Finite.exists_max fun i : ι ↦ ‖B.repr x i‖) i
+
+/-- For any `K`-basis of `L`, we have `B.norm 0 = 0`. -/
+protected theorem norm_zero : B.norm 0 = 0 := by
+  simp [norm, map_zero, Pi.zero_apply, norm_zero]
+
+/-- For any `K`-basis of `L`, and any `x : L`, we have `B.norm (-x) = B.norm x`. -/
+protected theorem norm_neg (x : L) : B.norm (-x) = B.norm x := by
+  simp [norm, map_neg,  Pi.neg_apply, _root_.norm_neg]
+
+variable {B}
+
+/-- For any `K`-basis of `L`, `B.norm` extends the norm on `K`. -/
+theorem norm_extends {i : ι} (hBi : B i = (1 : L)) :
+    Extends_map_algebraMap (fun x : K ↦ ‖x‖) B.norm := fun k ↦ by
+  classical
+  · by_cases hk : k = 0
+    · simp [hk, map_zero, norm_zero, Basis.norm_zero]
+    · simp only [norm, repr_algebraMap _ hBi]
+      have h_max : i = Classical.choose
+          (Finite.exists_max fun j : ι ↦ ‖(fun n : ι ↦ if i = n then k else 0) j‖) := by
+        by_contra h
+        have h_max' := Classical.choose_spec
+          (Finite.exists_max fun j : ι ↦ ‖(fun n : ι ↦ if i = n then k else 0) j‖)
+        simp only at h_max'
+        rw [if_neg h] at h_max'
+        specialize h_max' i
+        rw [if_pos rfl, norm_zero, norm_le_zero_iff] at h_max'
+        exact hk h_max'
+      rw [if_pos h_max]
+
+/-- For any `K`-basis of `L`, if the norm on `K` is nonarchimedean, then so is `B.norm`. -/
+theorem norm_isNonarchimedean (hna : IsNonarchimedean (Norm.norm : K → ℝ)) :
+    IsNonarchimedean B.norm := fun x y ↦ by
+  simp only [Basis.norm]
+  set ixy := Classical.choose (Finite.exists_max fun i : ι ↦ ‖B.repr (x + y) i‖)
+  have hxy : ‖B.repr (x + y) ixy‖ ≤ max ‖B.repr x ixy‖ ‖B.repr y ixy‖ := by
+    rw [LinearEquiv.map_add, Finsupp.coe_add, Pi.add_apply]; exact hna _ _
+  have hix := Classical.choose_spec (Finite.exists_max fun i : ι ↦ ‖B.repr x i‖)
+  have hiy := Classical.choose_spec (Finite.exists_max fun i : ι ↦ ‖B.repr y i‖)
+  rcases le_max_iff.mp hxy with (hx | hy)
+  · exact le_max_of_le_left (le_trans hx (hix ixy))
+  · exact le_max_of_le_right (le_trans hy (hiy ixy))
```

Comment:
```suggestion
theorem norm_isNonarchimedean (hna : IsNonarchimedean (Norm.norm : K → ℝ)) :
    IsNonarchimedean B.norm := fun x y ↦ by
  simp only [Basis.norm]
  obtain ⟨i, -, hi⟩ := exists_mem_eq_sup' univ_nonempty (fun i ↦ ‖(B.repr (x + y)) i‖)
  specialize hna ((B.repr x) i) ((B.repr y) i)
  aesop
```
(etc etc)

I know this is a very disruptive suggestion after all the work you've put in, but it looks like it makes the proofs much shorter. What do you think?